### PR TITLE
Refactor the preferences into a proper singleton.

### PR DESCRIPTION
**Pull request by Pentarctagon**
The current preferences handling is a mess:
* it's essentially a global config object that anything can modify in any way the caller wants, which is managed across multiple source files which have their own oddities and interdependencies.
* the general preferences has its own bit of SDL event handling and while I get the idea behind `events::sdl_handler` there's no reason to have SDL events handled in the preferences instead of just calling the relevant preferences setter for each event when it happens.
* the general preferences is where most of the preferences are handled and has its `base_manager` struct, which is part of the `manager` struct in the game preferences, which is then implicitly initialized as part of game_launcher's constructor.
* the editor preferences are the only preferences in a sub-namespace `preferences::editor` while all other preferences are just in the `preferences` namespace.
* the display, editor, and lobby preferences are all dependent on including the game preferences, the credentials are dependent on including the general preferences (but not the game preferences), the game preferences are dependent on including the general preferences, and the advanced preferences are entirely their own thing which is dependent on none of the other preference functionality and manages its own singleton.
* nothing checks whether the preferences file has actually been loaded before allowing values to be read from or written to the preferences config - if you attempt to get a value too early in wesnoth's initialization it will silently just give you whatever the default value for that preference happens to be.

With this there is instead a single access point (with exceptions handled via friend functions/classes), all predefined preferences are accessed via their own setter/getter, and all mainline preferences are defined in a single file (preference_list.hpp) so it's easily findable what preferences exist and where they're used. Having the list of all mainline preferences listed out also allows the lua preferences API to provide that full list rather than just the list of the preferences that have been set so far. Also it now checks for whether the location of the preferences file is known before attempting to load the preferences file and asserts if someone attempts to use the preferences too early.

---

@CelticMinstrel I added you as a reviewer for the lua_preferences changes.

### CODE REVIEW

**Comment by Vultraz** [comment]
I do wonder if it’s worth having setters and getters for every single preference. Would it doable to just publicly expose a static `operator[]` and let the callers access the desired pref as needed?

**Comment by Pentarctagon** [comment]
One of the main points is to limit how it can be accessed, so it's easy to know what exists and what's using a particular attribute.

**Comment by Vultraz** [comment]
Right, but since you added the scoped enum, that means only defined preferences can be accessed.

**Comment by Pentarctagon** [comment]
The MP alert preferences aren't part of the enum since there's some different handling based on the type (notif vs sound vs lobby). Different preferences also have different defaults as well as preference-specific logic.

File: src/preferences/preferences.hpp [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+		std::string default_dir();
+
+		bool draw_terrain_codes();
+		void set_draw_terrain_codes(bool value);
+
+		bool draw_hex_coordinates();
+		void set_draw_hex_coordinates(bool value);
+
+		bool draw_num_of_bitmaps();
+		void set_draw_num_of_bitmaps(bool value);
```
Comment (stevecotton) [inline src/preferences/preferences.hpp:521]: I accept the point about having a separate `preferences::editor::` namespace being different to the other preferences, but putting them all into a flat list like this loses all hints that these are settings that are only used in the editor.

I also think it would be a bad idea to unify the editor's persistent-across-restarts `draw_terrain_codes()` with the in-game but temporary `:draw_terrain_codes` commands.

File: src/preferences/preferences.hpp
Comment (stevecotton) [inline src/preferences/preferences.hpp:1]: Most of this file seems to have no namespace

File: src/preferences/preferences.hpp [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
+			// for boost: the userdata folders don't get initialized and the preferences aren't used for anything, so skip the check here
+			static bool called_before_init = !(filesystem::base_name(filesystem::get_exe_path()).find("boost") == std::string::npos && !filesystem::is_userdata_initialized());
+			assert(called_before_init);
+
+			static prefs prefs_manager;
+			return prefs_manager;
+		}
+
+		prefs();
+		~prefs();
```
Comment (stevecotton) [inline src/preferences/preferences.hpp:213]: Needs the other rule-of-five methods

File: src/preferences/preferences.hpp [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+			// for wesnoth: checks that the userdata folders have been initialized/found since if it hasn't been then it's too soon to be getting any values from it
+			// for boost: the userdata folders don't get initialized and the preferences aren't used for anything, so skip the check here
+			static bool called_before_init = !(filesystem::base_name(filesystem::get_exe_path()).find("boost") == std::string::npos && !filesystem::is_userdata_initialized());
+			assert(called_before_init);
+
+			static prefs prefs_manager;
+			return prefs_manager;
+		}
+
+		prefs();
```
Comment (stevecotton) [inline src/preferences/preferences.hpp:212]: The constructor should be private

File: src/events.cpp [hunk 4]
```diff
@@ -615,10 +616,11 @@ void pump()
 			// Here we can trigger any watchers for resize events.
 			// Video settings such as game_canvas_size() will be correct.
 			case SDL_WINDOWEVENT_RESIZED:
-				LOG_DP << "events/RESIZED "
+				LOG_DP << "events/RESIZED"
```
Comment (CelticMinstrel) [inline src/events.cpp:619]: Why did you remove a space here and on line 610?

**Comment by CelticMinstrel** [comment]
The changes to the Lua preferences API look fine to me.

Personally, I don't think it's good to merge all the preferences into a single file. Obviously the _headers_ have to be merged, since a singleton's definition can't really be be split across files. But the source files could have remained separate.

I'm also not too happy about the sound preference templates being inlined in the preferences dialog, as it's doing basically the exact same thing 4 times. I think I get why that was done, but was there really no way to keep them as separate functions, like using a pointer to member or a lambda or something?

**Comment by gfgtdf** [comment]
why are there so many ci fixes in this ? Are they in any way related ?

File: src/preferences/preferences.hpp (hunk 1 above)
Comment (stevecotton) [reply src/preferences/preferences.hpp:521]: There's a console command `:show_terrain_codes` (my mistake to call it `:draw_terrain_codes`), which is similar to what the editor does, but isn't saved between sessions. I meant it would be bad to have the editor's preference also turn it on or off during normal gameplay.

Comment (Pentarctagon) [reply src/preferences/preferences.hpp:521]: I don't think this PR changes that though?

File: src/preferences/preferences.hpp
Comment (Pentarctagon) [reply src/preferences/preferences.hpp:1]: Added them to the `preferences` namespace.

File: src/preferences/preferences.hpp (hunk 2 above)
Comment (Pentarctagon) [reply src/preferences/preferences.hpp:215]: Added.

File: src/preferences/preferences.hpp (hunk 3 above)
Comment (Pentarctagon) [reply src/preferences/preferences.hpp]: Updated.

File: src/events.cpp (hunk 4 above)
Comment (Pentarctagon) [reply src/events.cpp]: Updated.

**Comment by Pentarctagon** [comment]
@soliton- the credentials file now has its permissions set the same as the preferences file does.

@CelticMinstrel I've switch the preferences dialog sound options setup to the templated version.

**Comment by CelticMinstrel** [comment]
Cool. I still think it makes more sense to keep the various preferences source files separate though.

**Comment by Pentarctagon** [comment]
Having multiple cpp files for a single class is pretty unusual isn't it?

**Comment by Pentarctagon** [comment]
I'd prefer to instead do one or more of:
* separating the methods in the hpp+cpp files by area
* adding prefixes to the getter+setter methods
* adding doxygen comments

Aside from the editor preferences which were put in their own sub-namespace, the current preferences don't gain anything from being in separate files either - the rest are all called as just being in the `preferences` namespace regardless of which file they're defined in.

**Comment by CelticMinstrel** [comment]
> Having multiple cpp files for a single class is pretty unusual isn't it?

I don't think it's all that unusual for larger classes? And it allows being narrower about the includes, like not including crypto stuff outside the credentials file.

**Comment by Pentarctagon** [comment]
The crypto includes wouldn't have much of an impact would they, given they're in the cpp rather than the hpp file?

**Comment by CelticMinstrel** [comment]
I'm not really sure. Anything you include can potentially have an impact if it happens to define macros that conflict with the rest of the code (which windows.h commonly does), but I'm not sure if there's anything like that in these crypto headers.

**Comment by Pentarctagon** [comment]
Preferences that can probably be removed:
* custom_command - doesn't seem possible to actually use?
* editor_default_dir - has no setter

**Comment by Pentarctagon** [comment]
I think this PR is good to go. I'll wait a week to see if there are any additional review comments, but if there's nothing remaining to be done after a week I plan to merge this.

**Comment by CelticMinstrel** [comment]
> custom_command - doesn't seem possible to actually use?

What does that do?

**Comment by stevecotton** [comment]
Not sure what I'd use it for, but it works.

1. bind a hotkey for it
2. enter `:custom show_terrain_codes;show_num_of_bitmaps`
3. the hotkey now toggles the terrain codes and graphics count on and off

**Comment by Pentarctagon** [comment]
Huh, alright then.

**Comment by CelticMinstrel** [comment]
That seems kinda useful, but the interface to set it is kinda awful…

File: src/preferences/preferences.hpp [hunk 5]
```diff
@@ -0,0 +1,873 @@
...
+	}
+	static std::string get(const std::string& pref, const std::string& def)
+	{
+		return prefs::get().get(pref, def);
+	}
+};
+
+struct migrate_version_selection_friend
+{
+	friend class migrate_version_selection;
```
[the remark this answers is not in the document]
Comment (Pentarctagon) [reply src/preferences/preferences.hpp]: Updated.

File: src/preferences/preferences.cpp [hunk 6]
```diff
@@ -0,0 +1,2805 @@
...
+	}
+}
+
+prefs::~prefs()
+{
+	try {
+		if(!no_preferences_save_) {
+			write_preferences();
+		}
+	} catch (...) {}
```
[the remark this answers is not in the document]
Comment (Pentarctagon) [reply src/preferences/preferences.cpp]: Fixed - it ended up in the order I had moved things over from the old implementation but yeah, saving should be done last. Also added a log statement so on non-Windows platforms at least it'll say why it failed to save.
